### PR TITLE
[CS2113T-W09-1] Finux

### PR DESCRIPTION
**Pull request by tzexern**
Finux is a CLI Style application that allows the user to make better financial decisions based on the information recorded in the app. If you’re familiar with the CLI command interface, you will enjoy the benefit of speeding up your financial management rather than using the traditional management system.

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,34 +1,630 @@
...
+1. **Import the project as a Gradle Project.**
+   1. Click on `Import Project` and locate the `build.gradle` file and select it. Click `OK`.
+   1. If prompted, choose to `Open as Project` (not `Open as File`).
+   1. Click `OK` to accept the default settings but do ensure that the selected version of `Gradle JVM`
+      matches the JDK being used for the project.
+   1. Wait for the importing process to finish (it could take a few minutes).
+1. **Verify the setup.**
+   1. After importing successfully, locate the `src/main/java/seedu.duke/Duke.java` file, right click it, 
+      and choose `Run...`. If the setup is correct, you should see the following:
+   ![Login_Page](img/LoginPageExampleOutput.png)
```
Comment (8kdesign) [inline docs/DeveloperGuide.md:99]: Perhaps the unused space on the right side can be cropped out so that you can have more space to make the image bigger.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,34 +1,630 @@
...
+Before starting to write any code, we recommend that you have a look at Finux's overall design by reading
+[Finux's Architecture](#31-architecture) section.
+
+## 3. Design
+
+---
+
+### 3.1 Architecture
+
+![ArchitectureDiagram](img/ArchitectureDiagram.png)\
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md:129]: Should the arrows coming out of the "Finux" box be connected to some of the other boxes?

Comment (8kdesign) [inline docs/DeveloperGuide.md:129]: Perhaps the color of the light blue arrow can be changed. It is a bit hard to see at the moment.

File: src/main/java/seedu/duke/Finux.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * Starts the main application.
+     */
+    private void start() {
+        try {
+            ui = new Ui();
+            storage = new Storage();
+            parserHandler = new ParserHandler();
+            commandHandler = new CommandHandler();
+            storage.loadFile();
```
Comment (aliciatay-zls) [inline src/main/java/seedu/duke/Finux.java:52]: image


Was the call to loadFile() of the Finux class intentionally omitted?
Should "ArrayList" be "ArrayList//<Record//>" for the returning arrow after the call to getRecordListData() of Storage?

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,34 +1,630 @@
...
+
+The _**Architecture Diagram**_ above details the high-level design of the **Finux** Application.
+Brief explanations of the components involved are given below.
+
+The `Finux` component contains only one class `Finux`, It is responsible for,
+* At launch: Initializes the components in the correct sequence and attempts to load data from file.  
+* At shut down: Shuts down the components and invokes cleanup method where necessary.
+
+The rest of the Application consists of six components.
+* `Ui`: The user interface (Ui) of the App which handles all user input and Application output.
```
Comment (8kdesign) [inline docs/DeveloperGuide.md]: Is there a reason why both "App" and "Application" are used here? If they refer to the same thing, it might be confusing for the reader.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,34 +1,630 @@
...
+
+The Ui Component consists of a `Ui` class which handles all user input and system output.
+The Ui is only dependent on the `Duke` class and does not interact directly with other classes,
+ensuring a high level of cohesiveness, low level of coupling and separation of roles.
+
+The `Ui` component actively listens for:
+* the execution of commands to print the result of a `command`
+
+### 3.3 Parser Component
+![ParserHandlerClassDiagram](img/ParserHandlerClassDiagram.png) <br>
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md]: Looks a bit text-heavy, would it be possible to show only the most important methods of the 2 classes? I think it would be best to focus on the public methods, since they make up the API.

File: src/main/java/seedu/duke/Finux.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            String rawInput = ui.getUserInput();
+            ArrayList<String> parsedStringList = parserHandler.getParseInput(rawInput);
+            assert parsedStringList.size() != 0 : "Empty Parser Error";
+            Command command = commandHandler.parseCommand(parsedStringList, records);
+            if (command == null) {
+                continue;
+            }
+            command.execute(records, ui, storage, creditScoreMap);
+        }
+    }
```
Comment (aliciatay-zls) [inline src/main/java/seedu/duke/Finux.java]: image


Should the lifeline end at the cross for each of the classes?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,34 +1,630 @@
...
+2. `startExtraction` will check if the trimmed input starts with option. If yes, extract the option and 
+   remove the option from the trimmed input before calling `extractSubsequencePart`.
+3. `extractSubsequencePart` will check for the next option index and extract whatever is in between the start of the 
+   trimmed input to the start of the next option index. This is the argument tagged after the option. 
+   * Any leading or trailing white space of the argument field will be removed.
+   * If no argument is provided, the argument would be stored an empty string.
+
+   Afterward, the checking mechanism will loop until no valid next options are left in the input before calling
+   `extractFinalPart`.
+   * Valid next option format: `' <option> '` with 1 leading and trailing whitespace. e.g. `' -e '`
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md]: Not your fault, but it seems like the github pages theme breaks code formatting here. Maybe you could force the code block into a new line using `\`.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,34 +1,630 @@
...
+4. `extractFinalPart` will check if the last trimmed input ends with option. If yes, extract the option and 
+   add an empty string as the argument, else just add the last trimmed input to the ArrayList. 
+5. Finally, after the extraction to ArrayList<String> is complete, `extractFinalPart` will
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md]: Seems like the code blocks in this section are not being rendered properly. Also, should ArrayList be enclosed with code tags?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,34 +1,630 @@
...
+2. Checking of options in the start, end, and during the processing should be considered carefully with accordance 
+   to consideration 1.
+   * Starting option should be in the form of `'<option> '` with no leading whitespace and 1 trailing whitespace. 
+     e.g. `'-e '`
+   * Ending option should be in the form of `' <option>'` with 1 leading whitespace and no trailing whitespace.
+     e.g. `' -e'`
+   * During processing, next option should be in the form of `' <option> '` with 1 leading and trailing whitespace.
+     e.g. `' -e '`
+3. As rearrangement of options is allowed, option detection should cater to non-fixed option order. 
+   Apache Commons Lang, 3.11, providing the StringUtils class is used to cater to consideration 2.
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md]: Should code tags be used around StringUtils (same for the 3 bullet points below)?

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,34 +1,630 @@
...
+`Finux` to continue prompting for user input.
+
+Finally, if `parsedArguments[0]` cannot be mapped to any command, a `CommandException` is 
+thrown to `Finux` to handle.
+
+Not stated explicitly in the diagrams, when the `exit` command is entered, the `CommandHandler`
+sets the `isExit = true`, resulting in `Finux` proceeding to call `end()` to exit the Application.
+
+### 3.5 Command Component
+![CommandClassDiagram](img/CommandClassDiagram.png)
```
Comment (8kdesign) [inline docs/DeveloperGuide.md]: The image is too long so text appear too small. Would it be possible to shift some of the classes to a new line instead of arranging all of them in one line?

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,34 +1,630 @@
...
+5. As the ArrayList<String> is passed back to the main program and is being used by CommandHandler,
+   the argument field should compulsory and appended with empty string if empty to facilitate validations and option-argument
+   pairwise logic.
+6. getParseInput should always return a new ArrayList<String> per new input.
+
+
+### 3.4 CommandHandler Component
+![CommandHandlerClassDiagram](img/CommandHandlerClassDiagram.png)\
+_Figure 6: CommandHandler Class Diagram_\
+![CommandHandlerSequenceDiagram](img/CommandHandlerSequenceDiagram.png)\
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md:251]: Should the Finux lifeline also contain a activation bar here?

File: src/main/java/seedu/duke/Finux.java (hunk 3 above)
Comment (isaharon) [inline src/main/java/seedu/duke/Finux.java:52]: Should there be a call to the loadFile() method of Storage here? Also, the line for creditScoreMap seems to be omitted from the diagram as well. Was it probably intended?
Presentation1

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,34 +1,630 @@
...
+
+`list -l`
+
+To list expenses:
+
+`list -e`
+
+To list savings:
+
+`list -s`
```
Comment (8kdesign) [inline docs/DeveloperGuide.md]: For this part, perhaps you might want to remove it if you have nothing to share about its implementation.

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (brandonfoong) [inline docs/DeveloperGuide.md]: Would it be better to make this diagram bigger?

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,34 +1,630 @@
...
+
+`return -i <index of loan to return> -d <return date>`
+
+Upon marking the loan as returned, *creditscore* for the borrower who loaned the book will be updated and saved into `finux.txt`
+
+Example:
+`return -i 2 -d 2021-03-16`
+
+Output:
+Loan marked as returned: [L][2021-03-16] Loan to Tom [v]
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md]: Should this be in a code block?

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,34 +1,630 @@
...
+
+This section shows the design considerations taken when implementing the remove feature.
+
+Aspect: **When should the application validate the user input index**
+
+Since it is entirely possible that the index provided by the user can be invalid where the indices may be one of the
+following:
+* negative numbers
+* numbers that refer to non-existent records
+* non-numerics (e.g. alphabets, symbols, etc.)\
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md]: Seems like there's a small formatting bug here (extra `\`)

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,34 +1,630 @@
...
+of `Command` are the `AddCommand`, `CreditScoreCommand`, `ViewCommand`, etc...
+
+#### Design
+The only `abstract` method of `Command` is `execute(...)` where it is called everytime a `Command`
+object is created. Most of the input validation is done in the constructor of each Command object. 
+1. Firstly, arguments are checked for validity (if any):
+   1. Valid options: options only for each `Command`. E.g. `-a` for the amount `add` is valid.
+   2. No duplicate options: options are not repeated. E.g. `-a 200 -a 200` or `-l -l` is invalid.
+   3. No conflict options: mutually exclusive options, options that cannot be input at the same time.
+      E.g. ViewCommand implements `{-e | -l | -s}` options, `view -s -l` is considered a conflict of options.
```
Comment (8kdesign) [inline docs/DeveloperGuide.md]: Perhaps you might want to explain what {-e | -l | -s} means. I noticed it was explained in your user guide but not here.

File: src/main/java/seedu/duke/command/Command.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package seedu.duke.command;
+
+import seedu.duke.record.RecordList;
+import seedu.duke.storage.Storage;
+import seedu.duke.ui.Ui;
+
+public abstract class Command {
```
Comment (aliciatay-zls) [inline src/main/java/seedu/duke/command/Command.java:10]: image

Could this be hard to read/understand? Maybe just zoom in on one of the classes as an example e.g. AddCommand, since the purpose is just to describe that all these command classes have an attribute COMMAND_*?

File: docs/DeveloperGuide.md (hunk 13 above)
Comment (8kdesign) [inline docs/DeveloperGuide.md]: I think you forgot the inline code here.

File: src/main/java/seedu/duke/Finux.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+            System.exit(-1);
+        }
+    }
+
+    /**
+     * Loops the application until an EXIT command is parsed.
+     */
+    private void commandLooper() {
+        while (!commandHandler.isExit()) {
+            String rawInput = ui.getUserInput();
```
Comment (isaharon) [inline src/main/java/seedu/duke/Finux.java:68]: 1. Consider to add the return type `String` for this like the other parts below for consistency.
2. Method arguments need to be updated to match implementation. similarly for `execute()`.
Presentation1

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,34 +1,630 @@
...
+
+   Afterward, the checking mechanism will loop until no valid next options are left in the input before calling
+   `extractFinalPart`.
+   * Valid next option format: `' <option> '` with 1 leading and trailing whitespace. e.g. `' -e '`
+4. `extractFinalPart` will check if the last trimmed input ends with option. If yes, extract the option and 
+   add an empty string as the argument, else just add the last trimmed input to the ArrayList. 
+5. Finally, after the extraction to ArrayList<String> is complete, `extractFinalPart` will 
+   call `checkFirstBlock` for the final check to parse any `help` or `creditscore` in the first argument block.
+   
+#### Parser Component Design Consideration
```
Comment (brandonfoong) [inline docs/DeveloperGuide.md:221]: Would it be better to include the design considerations under implementation, rather than design? I believe that design is a high-level overview of what each component does, while implementation is for the finer details.

**Review by brandonfoong — COMMENTED** [review]
Overall, very good DG - most parts of the code are well-documented and the design choices are explained well. I have left some comments on the formatting and diagrams, but other than those, looks good!

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,34 +1,630 @@
...
+> * [Solution] Use a local based application.
+
+> * [Problem] Hassle to use different applications to keep track of various stuff.
+> * [Solution] Provide an integrated platform to record any expenses, loans, or saving.
+
+> * [Problem] Disorganization issues arising from keep multiple tracking applications.
+> * [Solution] Provide the option to view each category of records within an integrated platform.
+
+> * [Problem] Spending time to gauge whether to borrow money to a friend.
+> * [Solution] Provide a soft gauge through credit score indicators.
```
Comment (8kdesign) [inline docs/DeveloperGuide.md]: I like how you format it in terms of problem  -> solution.

File: src/main/java/seedu/duke/command/RemoveCommand.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+     * Prints a message containing the record that will be removed.
+     * Removes the record at the specified index.
+     *
+     * @param recordList is the recordList.
+     * @param ui      is the Ui object that interacts with the user.
+     * @param storage is the Storage object that reads and writes to the save file.
+     */
+    @Override
+    public void execute(RecordList recordList, Ui ui, Storage storage, CreditScoreMap creditScoreMap) {
+        Record currentRecord = recordList.getRecordAt(recordNumberInt);
```
Comment (aliciatay-zls) [inline src/main/java/seedu/duke/command/RemoveCommand.java:106]: image

Should there currentRecord:Record returned after the call to getRecordAt()?
Is it intentional that the parameters (recordList, creditScoreMap) in the saveData() call are omitted? It might be helpful to see these parameters.

**Review by 8kdesign — COMMENTED** [review]
Great job! Looking forward to seeing the completed version.

**Review by aliciatay-zls — COMMENTED** [review]
Diagrams are very detailed though sometimes overly so. Nice job!

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -1,34 +1,630 @@
...
+There is no set up required as the project comes with GitHub Actions config files, located in `.github/workflows`
+folder. When GitHub detects these files, it will run the CI for the project automatically at each push to the master
+branch or to any PR.
+
+#### 2.2.3 Learn the Design
+
+Before starting to write any code, we recommend that you have a look at Finux's overall design by reading
+[Finux's Architecture](#31-architecture) section.
+
+## 3. Design
```
Comment (isaharon) [inline docs/DeveloperGuide.md]: Design portion is quite comprehensive and done well. Keep up the good work!

File: docs/DeveloperGuide.md [hunk 22]
```diff
@@ -1,34 +1,630 @@
...
+
+The following sections below will provide more details of each component.
+
+### 3.2 UI Component
+
+![UIComponentDiagram](img/UIComponentDiagram.png)\
+*Figure 4: **Ui** Class Diagram*
+
+The Ui Component consists of a `Ui` class which handles all user input and system output.
+The Ui is only dependent on the `Duke` class and does not interact directly with other classes,
```
Comment (isaharon) [inline docs/DeveloperGuide.md]: Minor typo for `Duke`. Make sure to check for it e.g. `Setting Up` section in locating `Duke.java`.

File: docs/DeveloperGuide.md [hunk 23]
```diff
@@ -1,34 +1,630 @@
...
+
+> 💡 The lifeline all objects should end at the destroy marker (X) but due to a limitation of PlantUML,
+> the lifeline reaches the end of diagram.
+
+This sequence diagram follows suit after initialization in _Figure 2_.\
+This shows the main flow until the `exit` command is input by the user.
+
+The following sections below will provide more details of each component.
+
+### 3.2 UI Component
```
Comment (isaharon) [inline docs/DeveloperGuide.md:182]: Might want to add subheading `Description` for consistency with the other components.

**Review by isaharon — COMMENTED** [review]
Looks good! Quite readable from the start for the design part although some parts might be too wordy as compared to the rest.

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md:129]: It is intended, Finux initializes all the other components.
Thanks for asking!

File: docs/DeveloperGuide.md (hunk 11 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md:251]: Right, will add it in, thanks!

File: docs/DeveloperGuide.md (hunk 13 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md]: I think that was some sample placeholder for that section, we will renew it. Thanks!

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md:99]: Yes, thank you, will change it!

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md:129]: Sure no problem!

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md]: Thanks for pointing this out! Will change it.

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md]: PlantUML limitations, decided to split the diagram into 2 images.
Thanks!

File: docs/DeveloperGuide.md (hunk 12 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md]: I think that was some sample placeholder for that section, we will renew it. Thanks!

File: docs/DeveloperGuide.md (hunk 15 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md]: Thanks for pointing this out! Will move it over.

File: docs/DeveloperGuide.md (hunk 13 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md]: I think that was some sample placeholder for that section, we will renew it. Thanks!

File: src/main/java/seedu/duke/Finux.java (hunk 3 above)
Comment (marklowsk) [reply src/main/java/seedu/duke/Finux.java:52]: Thanks for pointing this out! We have changed it.

File: src/main/java/seedu/duke/Finux.java (hunk 6 above)
Comment (marklowsk) [reply src/main/java/seedu/duke/Finux.java]: PlantUML limitation, we have explained it in one of the sections.

File: src/main/java/seedu/duke/command/Command.java (hunk 16 above)
Comment (marklowsk) [reply src/main/java/seedu/duke/command/Command.java:10]: Thanks for pointing this out! PlantUML limitations, decided to split the diagram into 2 images.

File: src/main/java/seedu/duke/Finux.java (hunk 3 above)
Comment (marklowsk) [reply src/main/java/seedu/duke/Finux.java:52]: Thanks for pointing this out! We have changed it.

File: src/main/java/seedu/duke/Finux.java (hunk 17 above)
Comment (marklowsk) [reply src/main/java/seedu/duke/Finux.java:68]: Thanks for pointing this out! Will rectify the errors.
We are still in the process of standardizing our diagrams.

File: docs/DeveloperGuide.md (hunk 23 above)
Comment (marklowsk) [reply docs/DeveloperGuide.md:182]: hmmm. Will discuss with my teammates.

File: src/main/java/seedu/duke/command/RemoveCommand.java (hunk 20 above)
Comment (marklowsk) [reply src/main/java/seedu/duke/command/RemoveCommand.java:106]: Thanks for pointing this out! Will rectify the errors.
We are still in the process of standardizing our diagrams.
